lib/engine/game/g_1848/game.rb

```diff
@@ -658,6 +688,79 @@ def market_share_limit(corporation = nil)
 
           MARKET_SHARE_LIMIT
         end
+
+        def can_take_loan?(entity)
+          entity.corporation? &&
+            entity.loans.size < maximum_loans(entity) &&
+            @loans.any?
+        end
+
+        def take_loan(entity, loan)
+          raise GameError, "Cannot take more than #{maximum_loans(entity)} loans" unless can_take_loan?(entity)
+
+          old_price = entity.share_price
+          boe_old_price = @boe.share_price
+          name = entity.name
+          @log << "#{name} takes a loan and receives #{format_currency(loan.amount)}"
+          @boe.spend(loan.amount, entity)
+          loan_taken_stock_market_movement(entity)
+          log_share_price(entity, old_price)
+          log_share_price(@boe, boe_old_price)
+          entity.loans << loan
+          @boe.loans << loan
+          @loans.delete(loan)
+        end
+
+        def loan_taken_stock_market_movement(entity)
+          @stock_market.move_left(entity)
+          @stock_market.move_left(entity)
+          @stock_market.move_right(boe)
+        end
+
+        # routing logic
+        def check_distance(route, visits, _train = nil)
+          gauge_changes = edge_crossings(route)
+          modifier = route.train.name.include?('+') ? 1 : 0
+          modified_guage_changes = gauge_changes - modifier
+          visits += Array.new(modified_guage_changes) { Engine::Part::City.new('0') } if modified_guage_changes.positive?
+
+          super(route, visits, _train = nil)
+        end
+
+        def route_distance(route)
+          gauge_changes = edge_crossings(route)
+          modifier = route.train.name.include?('+') ? 1 : 0
+          distance = super
+          distance + gauge_changes - modifier
+        end
+
+        def edge_crossings(route)
+          sum = route.paths.sum do |path|
+            path.edges.sum do |edge|
+              edge_is_a_border(edge) ? 1 : 0
+            end
+          end
+          # edges are double counted
+          sum / 2
+        end
+
+        def edge_is_a_border(edge)
+          border_edges = []
```

Comment:
Does this all collapse down to

edge.hex.tile.borders.any? { |boarder| boarder.edge == edge.num }